src/AsmResolver.DotNet/Signatures/GenericContext.cs

```diff
@@ -62,14 +69,111 @@ public TypeSignature GetTypeArgument(GenericParameterSignature parameter)
                 GenericParameterType.Method => Method,
                 _ => throw new ArgumentOutOfRangeException()
             };
-            
+
             if (argumentProvider is null)
-                throw new ArgumentOutOfRangeException();
+                return parameter;
 
             if (parameter.Index >= 0 && parameter.Index < argumentProvider.TypeArguments.Count)
                 return argumentProvider.TypeArguments[parameter.Index];
 
             throw new ArgumentOutOfRangeException();
         }
+
+
+        /// <summary>
+        /// Gets type generic context from <see cref="TypeSpecification"/>.
+        /// </summary>
+        /// <param name="type">Type specification to get generic context from.</param>
+        /// <returns>Generic context.</returns>
+        public static GenericContext FromType(TypeSpecification type) => type.Signature switch
+        {
+            GenericInstanceTypeSignature typeSig => new GenericContext(typeSig, null),
+            _ => default
+        };
+
+        /// <summary>
+        /// Gets type generic context from <see cref="GenericInstanceTypeSignature"/>.
+        /// </summary>
+        /// <param name="type">Generic type signature to get generic context from.</param>
+        /// <returns>Generic context.</returns>
+        public static GenericContext FromType(GenericInstanceTypeSignature type) => new(type, null);
+
+        /// <summary>
+        /// Gets type generic context from <see cref="ITypeDescriptor"/>.
+        /// </summary>
+        /// <param name="type">Type to get generic context from.</param>
+        /// <returns>Generic context.</returns>
+        public static GenericContext FromType(ITypeDescriptor type) => type switch
+        {
+            TypeSpecification typeSpecification => FromType(typeSpecification),
+            GenericInstanceTypeSignature typeSig => FromType(typeSig),
+            _ => default
+        };
+
+        /// <summary>
+        /// Gets method and/or type generic context from <see cref="MethodSpecification"/>.
+        /// </summary>
+        /// <param name="method">Method specification to get generic context from.</param>
+        /// <returns>Generic context.</returns>
+        public static GenericContext FromMethod(MethodSpecification method) =>
+            (method.DeclaringType, method.Signature) switch
```

Comment:
This is some pretty hefty `switch` pattern matching abuse. Maybe consider normal if statements for readability?